module/mempool/stdmap/backdata/arraylinkedlist/entitiesLinkedList.go

```diff
@@ -0,0 +1,350 @@
+package arraylinkedlist
+
+import (
+	"math/rand"
+
+	"github.com/onflow/flow-go/model/flow"
+)
+
+type EjectionMode string
+
+const (
+	RandomEjection = EjectionMode("random-ejection")
+	LRUEjection    = EjectionMode("lru-ejection")
+)
+
+// number of entries we try to eject before giving up and ejecting LRU
+const maximumRandomTrials = 10
+
+// EIndex is data type representing an entity index in EntityDoubleLinkedList.
+type EIndex uint32
+
+// doubleLinkedListPointer represents a slice-based linked list pointer. Instead of pointing
+// to a memory address, this pointer points to a slice index.
+//
+// Note: an "undefined" (i.e., nil) notion for this doubleLinkedListPointer corresponds to the
+// value of uint32(0). Hence, legit "pointer" values start from uint32(1).
+// doubleLinkedListPointer also furnished with methods to convert a "pointer" value to
+// a slice index, and reverse.
+type doubleLinkedListPointer struct {
+	pointerValue uint32
+}
+
+// isUndefined returns true if this pointer is set to its zero value. An undefined
+// slice-based pointer is equivalent to a nil address-based one.
+func (d doubleLinkedListPointer) isUndefined() bool {
+	return d.pointerValue == uint32(0)
+}
+
+// setUndefined sets sliced-based pointer to its undefined (i.e., nil equivalent) value.
+func (d *doubleLinkedListPointer) setUndefined() {
+	d.pointerValue = uint32(0)
+}
+
+// sliceIndex returns the slice-index equivalent of the pointer.
+func (d doubleLinkedListPointer) sliceIndex() EIndex {
+	return EIndex(d.pointerValue) - 1
+}
+
+// setPointer converts the input slice-based index into a slice-based pointer and
+// sets the underlying pointer.
+func (d *doubleLinkedListPointer) setPointer(sliceIndex EIndex) {
+	d.pointerValue = uint32(sliceIndex + 1)
+}
+
+// doubleLinkedListNode represents a slice-based double linked list node that
+// consists of a next and previous pointer.
+type doubleLinkedListNode struct {
+	next doubleLinkedListPointer
+	prev doubleLinkedListPointer
+}
+
+// doubleLinkedList represents a double linked list by its head and tail pointers.
+type doubleLinkedList struct {
+	head doubleLinkedListPointer
+	tail doubleLinkedListPointer
+}
+
+func newDoubleLinkedList() *doubleLinkedList {
+	return &doubleLinkedList{
+		head: doubleLinkedListPointer{pointerValue: 0},
+		tail: doubleLinkedListPointer{pointerValue: 0},
+	}
+}
+
+// cachedEntity represents a cached entity that is maintained as a double linked list node.
+type cachedEntity struct {
+	owner  uint64
+	node   doubleLinkedListNode
+	id     flow.Identifier
+	entity flow.Entity
+}
```

Comment:
Here, `id` is the key, and `entity` is the value. It would be great if this could somehow be abstracted so that the API can operate on any caller supplied key and/or value struct supplied, as suggested here [1].

[1] https://www.notion.so/dapperlabs/Intro-to-heap-friendly-hero-structures-cd613aba6f4244f88e58fec1f222f476

Reply:
Issued up https://github.com/dapperlabs/flow-go/issues/6084 to revisit it later. Not sure if supporting generic `interface {}` would be a safe idea. To the best of my knowledge, the generic data structure in Flow codebase is the 
[`flow.Entity`](https://github.com/onflow/flow-go/blob/master/model/flow/entity.go), which is an `interface` (and _not_ a value type). An `Entity` is any type that implements `ID` and `CheckSum` methods. The reason is to the best of my knowledge, in our Flow codebase,  everything is purposefully meant to be keyed through a well-defined identifier of fixed size, which is equal in the length to the output of the collision-resistant hash function that we utilize (i.e., SHA3-256). This standard is realized through defining the `flow.Entity` that is identifiable through the `flow.Identifier` type. This makes all of our data structures self-certifying, which is a cardinal principle in designing blockchains, i.e., the _integrity_ of the obtained `entity` corresponding to the requested/retrieved `id`, can be checked by anyone through `entity.ID() == id`. 

So, unless we export HeroCache as a separate repository, or find a use-case, sticking to the Flow-defined generic type (i.e., `Entity`) seems the safest, and refactoring to Golang native generic type may jeopardize self-certifying correctness feature of our codebase.

Reply:
Thanks, commented in the issue :-)